Code/Core/Strings/AString.cpp

```diff
@@ -580,6 +580,215 @@ void AString::ToUpper()
     }
 }
 
+// JSONEscape
+//------------------------------------------------------------------------------
+void AString::JSONEscape()
+{
+    // First find how much extra space is needed to store escaped characters
+    uint32_t extraSpace = 0;
+    char * pos = m_Contents;
+    char * end = m_Contents + m_Length;
+    while ( pos < end )
+    {
+        char c = *pos;
+        if ( ( c == 0x08 ) || ( c == 0x09 ) || ( c == 0x0A ) || ( c == 0x0C ) || ( c == 0x0D ) || ( c == '"' ) || ( c == '\\') )
+        {
+            extraSpace += 1;
+        }
+        else if ( ( c >= 0 ) && ( c <= 0x1F ) )
+        {
+            extraSpace += 5;
+        }
+        ++pos;
+    }
+
+    if ( extraSpace == 0 )
+    {
+        return; // No characters need escaping
+    }
+
+    if ( m_Length + extraSpace > GetReserved() )
+    {
+        Grow( m_Length + extraSpace );
+    }
+
```

Comment:
I think this function would be much simpler if it built a temporary buffer for the output and swapped the strings. That would allow the loop to go forwards. As is, the grow already causes a temporary copy to occur, so it wouldn't be any more costly.